### PR TITLE
Integrate XRAI repository with FoXAI

### PR DESCRIPTION
**Pull request by adamwawrzynski**
## Description

Add implementation of XRAI algorithm.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

- [x] `pre-commit run`

## Checklist:

- [ ] I have updated all related `pyproject.toml` files.
- [x] I have updated lock files.
- [x] I have set an Assignee and Reviewers.
- [x] I have labeled the PR correctly.
- [x] I have performed a self-review of my own code.
- [ ] I have covered my code and changes with unit tests.
- [x] I have annotated types as extensively as possible.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] All code references and dependencies will work.

### CODE REVIEW

File: foxai/explainer/computer_vision/algorithm/xrai.py [hunk 1]
```diff
@@ -0,0 +1,692 @@
...
+    image_max = np.max(image_batch)
+    image_min = np.min(image_batch)
+    image_batch = (image_batch - image_min) / (image_max - image_min)
+    image_batch = image_batch * (value_range[1] - value_range[0]) + value_range[0]
+    color_channels: int = image_batch.shape[-1]
+    resized_image_list: List[np.ndarray] = []
+    for image in image_batch:
+        if resize_shape is not None:
+            # resize can only process single image with shape H x W x C
+            resized_image_list.append(
```
Comment (joannasendorek) [inline foxai/explainer/computer_vision/algorithm/xrai.py:50]: Do I understand correctly that resized_image_list will be empty by default? Meaning the result also will be empty by default (default resize_shape is None). Why is resize_shape Optional then if when not provided, the whole method doesn't make sense?

File: foxai/explainer/computer_vision/algorithm/xrai.py [hunk 2]
```diff
@@ -0,0 +1,687 @@
...
+    """
+    masks: List[np.ndarray] = []
+    for segment in segment_list:
+        for index in range(segment.min(), segment.max() + 1):
+            masks.append(segment == index)
+    return masks
+
+
+class XRAI:
+    """A CoreSaliency class that computes saliency masks using the XRAI method."""
```
Comment (joannasendorek) [inline foxai/explainer/computer_vision/algorithm/xrai.py:225]: This docstring seems off.

File: foxai/explainer/computer_vision/algorithm/xrai.py [hunk 3]
```diff
@@ -0,0 +1,687 @@
...
+                    additional_forward_args=call_model_args,
+                )
+            )
+
+        return grads
+
+    @staticmethod
+    def _validate_baselines(
+        input_data: torch.Tensor,
+        baselines: torch.Tensor,
```
Comment (joannasendorek) [inline foxai/explainer/computer_vision/algorithm/xrai.py:276]: Is it a BaselineType?

File: foxai/explainer/computer_vision/algorithm/xrai.py [hunk 4]
```diff
@@ -0,0 +1,687 @@
...
+                least important segment. If integer_segments is set to `False`,
+                the segments are returned as a boolean array, where the first
+                dimension has size `N`. The `[0, ...]` mask is the most important and the
+                `[N-1, ...]` mask is the least important. Defaults to `True`.
+
+        Returns:
+            Saliency heatmap `np.ndarray` and masks or an integer image with
+                area ranks depending on the parameter integer_segments.
+        """
+        attributes_list: List[np.ndarray] = []
```
Comment (joannasendorek) [inline foxai/explainer/computer_vision/algorithm/xrai.py:494]: AttributionsType?

File: foxai/explainer/computer_vision/algorithm/xrai.py [hunk 5]
```diff
@@ -0,0 +1,687 @@
...
+        area_perc_th: float,
+        min_pixel_diff: int,
+        integer_segments: bool,
+        attribute_sample: np.ndarray,
+        segments: List[np.ndarray],
+    ) -> Tuple[np.ndarray, List[np.ndarray]]:
+        """Run XRAI saliency given single attribution and segment.
+
+        Args:
+            attributes: Source attribution for XRAI.
```
Comment (joannasendorek) [inline foxai/explainer/computer_vision/algorithm/xrai.py:529]: No such param

File: foxai/explainer/computer_vision/algorithm/xrai.py [hunk 6]
```diff
@@ -0,0 +1,687 @@
...
+            integer_segment_mask = np.zeros(shape=attribute_sample.shape, dtype=int)
+            for i, mask in enumerate(masks_trace_array_list):
+                integer_segment_mask[mask] = i + 1
+
+            final_mask_trace = [integer_segment_mask]
+        return output_attr, final_mask_trace
+
+
+class XRAICVExplainer(Explainer):
+    """XRAI algorithm explainer."""
```
Comment (joannasendorek) [inline foxai/explainer/computer_vision/algorithm/xrai.py:622]: Perhaps some link would be helpful

File: foxai/explainer/computer_vision/algorithm/xrai.py [hunk 7]
```diff
@@ -0,0 +1,687 @@
...
+
+    def calculate_features(
+        self,
+        model: ModelType,
+        input_data: torch.Tensor,
+        pred_label_idx: TargetType = None,
```
Comment (joannasendorek) [inline foxai/explainer/computer_vision/algorithm/xrai.py:628]: Please use Optional[X] if None is allowed.

File: tests/test_array_utils.py [hunk 8]
```diff
@@ -328,26 +332,893 @@ def test_transpose_color_in_array_with_batch_size_should_work() -> None:
...
+                    [29],
+                    [30],
+                ],
+                [
+                    [31],
+                    [32],
+                    [33],
+                    [34],
+                    [35],
+                    [36],
```
Comment (joannasendorek) [inline tests/test_array_utils.py:965]: Is it automatic formatting? It's hardly readable to me.

Comment (adamwawrzynski) [reply tests/test_array_utils.py:965]: Yes, it is automatic.

Comment (joannasendorek) [reply tests/test_array_utils.py:965]: Let it be then.

File: foxai/explainer/computer_vision/algorithm/xrai.py (hunk 7 above)
Comment (adamwawrzynski) [reply foxai/explainer/computer_vision/algorithm/xrai.py:628]: I see that all explanation methods have this issue. I will create another PR with typing-related issues.

File: foxai/explainer/computer_vision/algorithm/xrai.py (hunk 4 above)
Comment (adamwawrzynski) [reply foxai/explainer/computer_vision/algorithm/xrai.py:494]: `AttributionsType` is defined as `torch.Tensor`. All types have to be coherent in public methods/API but internally it makes more sense be less restrictive.

File: foxai/explainer/computer_vision/algorithm/xrai.py (hunk 3 above)
Comment (adamwawrzynski) [reply foxai/explainer/computer_vision/algorithm/xrai.py]: Fixed: https://github.com/softwaremill/FoXAI/pull/99/commits/49a269c31fe09c365852daba10e1a93d5fe9850a.

File: foxai/explainer/computer_vision/algorithm/xrai.py (hunk 1 above)
Comment (adamwawrzynski) [reply foxai/explainer/computer_vision/algorithm/xrai.py]: Fixed by splitting this function into two: https://github.com/softwaremill/FoXAI/pull/99/commits/d13394acecd5f292cb4c1d59eb03f18d16b93b4c.

File: foxai/explainer/computer_vision/algorithm/xrai.py (hunk 2 above)
Comment (adamwawrzynski) [reply foxai/explainer/computer_vision/algorithm/xrai.py]: Fixed: https://github.com/softwaremill/FoXAI/pull/99/commits/49a269c31fe09c365852daba10e1a93d5fe9850a.

File: foxai/explainer/computer_vision/algorithm/xrai.py (hunk 5 above)
Comment (adamwawrzynski) [reply foxai/explainer/computer_vision/algorithm/xrai.py]: Fixed: https://github.com/softwaremill/FoXAI/pull/99/commits/49a269c31fe09c365852daba10e1a93d5fe9850a.

File: foxai/explainer/computer_vision/algorithm/xrai.py (hunk 6 above)
Comment (adamwawrzynski) [reply foxai/explainer/computer_vision/algorithm/xrai.py]: Fixed: https://github.com/softwaremill/FoXAI/pull/99/commits/49a269c31fe09c365852daba10e1a93d5fe9850a.

File: foxai/explainer/computer_vision/algorithm/xrai.py (hunk 1 above)
Comment (joannasendorek) [reply foxai/explainer/computer_vision/algorithm/xrai.py]: I really like this change, it is much better.

**Review by joannasendorek — APPROVED** [review]
Looks great!

**Pull request merged**